### PR TITLE
[WIP] Gazebo Optimisation (Brandon)

### PR DESCRIPTION
**Pull request by Jade-Fountain**
Implements both a connection to [Gazebo](github.com/NUbots/Gazebo) plugin and also NSGA2 Multiobjective optimiser for scripts

### CODE REVIEW

**Comment by Bidski** [comment]
> Not a full review yet, I’ll do that later. But for starters, delete any commented out code and replace any cout lines with nuclear logs of the correct log level (info/warn/error etc)

Re-review please @TrentHouliston

File: module/support/optimisation/NSGA2Optimiser/src/random.cpp [hunk 1]
```diff
@@ -0,0 +1,32 @@
+#include "random.h"
```
Comment (TrentHouliston) [inline module/support/optimisation/NSGA2Optimiser/src/random.cpp]: in general, I'm not reviewing any of the NSGA stuff... But all this is in the standard library and why are we using boost for this!!

File: nuclear/b.py [hunk 2]
```diff
@@ -54,7 +54,7 @@
     # Remove lines that are comments
     if len(l) > 0 and not l.startswith("//") and not l.startswith("#"):
         # Extract our variable name from our values
-        g = re.match(r"([a-zA-Z_$][a-zA-Z_.$0-9-]*):(\w+)=(.*)", l).groups()
+        g = re.match(r"([a-zA-Z_$][a-zA-Z_ .$0-9-]*):(\w+)=(.*)", l).groups()
```
Comment (TrentHouliston) [inline nuclear/b.py]: There are really spaces in cmake variable names?
That's terrible

File: puppet/manifests/nubots.pp [hunk 3]
```diff
@@ -197,7 +421,8 @@
         installer { "${archive}":
           archs       => $archs,
           creates     => $params['creates'],
-          require     => delete_undef_values(flatten([ Archive["${archive}"], $params['require'], Class['installer::prerequisites'], Class['build_tools'], ])),
+          # Gazebo depends on some of our installed libraries
```
Comment (TrentHouliston) [inline puppet/manifests/nubots.pp]: What is this line?

File: puppet/manifests/nubots.pp [hunk 4]
```diff
@@ -307,6 +529,108 @@
       before  => [ Installer['aravis'], ],
     }
 
+    file { "${prefix}/${arch}/src/dart.patch":
```
Comment (TrentHouliston) [inline puppet/manifests/nubots.pp]: Why are we patching stuff in our puppet file?

File: puppet/manifests/nubots.pp [hunk 5]
```diff
@@ -339,6 +663,10 @@
 set(CMAKE_CXX_FLAGS \"\${CMAKE_CXX_FLAGS} ${compile_params}\" CACHE STRING \"\")
 
 set(PLATFORM \"${arch}\" CACHE STRING \"The platform to build for.\" FORCE)
+
+# Set cmake up to look in our pkgconfig directories
```
Comment (TrentHouliston) [inline puppet/manifests/nubots.pp]: Should this be in a separate PR?

File: puppet/modules/build_tools/manifests/init.pp [hunk 6]
```diff
@@ -76,6 +76,34 @@
...
+  package { 'libyaml-dev': ensure => latest, }
+  package { 'libtbb-dev': ensure => latest, }
+  package { 'freeglut3-dev': ensure => latest, }
+  package { 'libogre-1.9-dev': ensure => latest, }
+  package { 'libtar-dev': ensure => latest, }
+  package { 'libqwt-qt5-dev': ensure => latest, }
+  package { 'libqt5core5a': ensure => latest, }
+  package { 'libqt5opengl5-dev': ensure => latest, }
+  package { 'libqt5test5': ensure => latest, }
+  package { 'libqt5opengl5': ensure => latest, }
```
Comment (TrentHouliston) [inline puppet/modules/build_tools/manifests/init.pp]: We need qt to build something in a headless server?

File: puppet/modules/build_tools/manifests/init.pp [hunk 7]
```diff
@@ -76,6 +76,34 @@
...
+  package { 'libgts-dev': ensure => latest, }
+  package { 'libavdevice-dev': ensure => latest, }
+  package { 'libavformat-dev': ensure => latest, }
+  package { 'libavcodec-dev': ensure => latest, }
+  package { 'libswscale-dev': ensure => latest, }
+  package { 'libavutil-dev': ensure => latest, }
+  package { 'libzip-dev': ensure => latest, }
+  package { 'libjsoncpp-dev': ensure => latest, }
+  package { 'libcurl4-openssl-dev': ensure => latest, }
+  package { 'libyaml-dev': ensure => latest, }
```
Comment (TrentHouliston) [inline puppet/modules/build_tools/manifests/init.pp]: Don't we already build some of these?

File: puppet/modules/installer/manifests/init.pp [hunk 8]
```diff
@@ -224,7 +226,7 @@
           onlyif      => "test \"${method}\" = \"autotools\" ",
           # Sometimes, autogen.sh will automatically run configure
           command     => "${prebuild_cmd} &&
-                          if [ -e \"autogen.sh\" ]; then NOCONFIGURE=1 ./autogen.sh; fi &&
+                          if [ -e \"autogen.sh\" ]; then NOCONFIGURE=1 ./autogen.sh --no-configure; fi &&
```
Comment (TrentHouliston) [inline puppet/modules/installer/manifests/init.pp]: What is this for?

File: shared/message/motion/ExecuteKick.proto [hunk 9]
```diff
@@ -0,0 +1,31 @@
...
+
+package message.motion;
+
+/**
+ * TODO document
+ *
+ * @author Trent Houliston
+ * @author Brendan Annable
+ */
+message ExecuteKick {
```
Comment (TrentHouliston) [inline shared/message/motion/ExecuteKick.proto]: What even is this proto file for?

File: shared/message/support/Gazebo/Gazebo.proto [hunk 10]
```diff
@@ -0,0 +1,14 @@
+syntax = "proto3";
+
+package message.support.gazebo;
+
+/// \brief String message
+
+message StringMsg {
```
Comment (TrentHouliston) [inline shared/message/support/Gazebo/Gazebo.proto]: Is this the string we send to/from gazebo? Can't we just send real messages?

File: shared/message/support/Gazebo/GazeboBallLocation.proto [hunk 11]
```diff
@@ -0,0 +1,16 @@
+syntax = "proto3";
+
+package message.support.Gazebo;
```
Comment (TrentHouliston) [inline shared/message/support/Gazebo/GazeboBallLocation.proto]: captial letter in package name?

File: shared/message/support/Gazebo/GazeboBallLocation.proto [hunk 12]
```diff
@@ -0,0 +1,16 @@
...
+
+message GazeboBallLocation {
+
+    double x = 1;
```
Comment (TrentHouliston) [inline shared/message/support/Gazebo/GazeboBallLocation.proto]: We have vec3s for this. Also, this should be a vector of the form rABc so we know what space it's measured in etc

File: shared/message/support/Gazebo/GazeboBallLocation.proto [hunk 13]
```diff
@@ -0,0 +1,16 @@
...
+    double y = 2;
+    double z = 3;
+
+    double velx = 4;
+    double vely = 5;
+    double velz = 6;
+
+    string sourceName = 7;
```
Comment (TrentHouliston) [inline shared/message/support/Gazebo/GazeboBallLocation.proto]: What is this variable, also why is it camel case?

File: shared/message/support/Gazebo/GazeboRobotLocation.proto [hunk 14]
```diff
@@ -0,0 +1,11 @@
+syntax = "proto3";
+
+package message.support.Gazebo;
+
+message GazeboRobotLocation {
+    double x = 1;
```
Comment (TrentHouliston) [inline shared/message/support/Gazebo/GazeboRobotLocation.proto]: vec3 of the form rABc

Also, can't we get orientation so we can get Htw?

File: shared/message/support/Gazebo/GazeboWorldCtrl.proto [hunk 15]
```diff
@@ -0,0 +1,7 @@
+syntax = "proto3";
+
+package message.support.Gazebo;
```
Comment (TrentHouliston) [inline shared/message/support/Gazebo/GazeboWorldCtrl.proto]: So sometimes we get real protobufs?

File: shared/message/support/Gazebo/GazeboWorldCtrl.proto [hunk 16]
```diff
@@ -0,0 +1,7 @@
...
+
+message GazeboWorldCtrl {
```
Comment (TrentHouliston) [inline shared/message/support/Gazebo/GazeboWorldCtrl.proto]: Sending opaque strings is a terrible way to communicate. Also abbreviating Control to Ctrl is not that much of a saving

File: shared/message/support/Gazebo/GazeboWorldStatus.proto [hunk 17]
```diff
@@ -0,0 +1,13 @@
+syntax = "proto3";
+
+package message.support.Gazebo;
+
+message GazeboWorldStatus {
+    double simTime   = 1;
```
Comment (TrentHouliston) [inline shared/message/support/Gazebo/GazeboWorldStatus.proto]: should be `snake_case`

File: shared/utility/clock/CustomClock.cpp [hunk 18]
```diff
@@ -0,0 +1,23 @@
+#include "CustomClock.hpp"
+#include <chrono>
+
+namespace utility {
+namespace clock {
+    double custom_rtf;  // real time factor
+    std::chrono::steady_clock::time_point lastUpdate;
```
Comment (TrentHouliston) [inline shared/utility/clock/CustomClock.cpp]: inconsistent variable case again

File: install_gazebo.sh [hunk 19]
```diff
@@ -0,0 +1,410 @@
+#! /bin/bash
```
[the remark this answers is not in the document]
Comment (Bidski) [reply install_gazebo.sh]: To simplify installing gazebo externally to the VM so you can use its fancy GUI (assuming you are running a linux-like OS that may not have gazebo in the system package manager).

File: cmake/Modules/SharedLibraries.cmake [hunk 20]
```diff
@@ -2,7 +2,8 @@
 FIND_PACKAGE(Armadillo REQUIRED)
 FIND_PACKAGE(LibGFortran REQUIRED)
 FIND_PACKAGE(OpenBLAS)
-FIND_PACKAGE(Protobuf REQUIRED)
+FIND_PACKAGE(ignition-cmake0 REQUIRED)
```
[the remark this answers is not in the document]
Comment (Bidski) [reply cmake/Modules/SharedLibraries.cmake]: you think I like it?

File: module/motion/OldWalkEngine/data/config/nubotsvm/OldWalkEngine.yaml [hunk 21]
```diff
@@ -0,0 +1,167 @@
+stance:
```
[the remark this answers is not in the document]
Comment (Bidski) [reply module/motion/OldWalkEngine/data/config/nubotsvm/OldWalkEngine.yaml]: was probably a result of walk training with gazebo in the vm?

File: module/motion/ScriptEngine/data/scripts/nubotsvm/KickLeft.yaml [hunk 22]
```diff
@@ -0,0 +1,444 @@
+- duration: 1000
```
[the remark this answers is not in the document]
Comment (Bidski) [reply module/motion/ScriptEngine/data/scripts/nubotsvm/KickLeft.yaml]: maybe these are the fabled gazebo kicks that we should have used at robocup?

File: module/motion/IKKick/src/IKKick.cpp [hunk 23]
```diff
@@ -24,6 +24,7 @@
 #include "message/behaviour/KickPlan.h"
 #include "message/behaviour/ServoCommand.h"
 #include "message/input/Sensors.h"
+#include "message/motion/ExecuteKick.h"
```
[the remark this answers is not in the document]
Comment (Bidski) [reply module/motion/IKKick/src/IKKick.cpp]: it is used

File: module/support/Gazebo/src/Gazebo.cpp [hunk 24]
```diff
@@ -0,0 +1,444 @@
...
+                    std::string sourceName;
+                    std::getline(ss, sourceName);
+                    remotes.insert(sourceName);
+
+                    if (remotes.find(sourceName) == remotes.begin()) {
+                        // Servos
+                        for (int i = 0; i < 20; ++i) {
+                            // Get a reference to the servo we are populating
+                            DarwinSensors::Servo& servo = utility::platform::darwin::getDarwinServo(i, *sensors);
+                            std::getline(ss, line);
```
[the remark this answers is not in the document]
Comment (Bidski) [reply module/support/Gazebo/src/Gazebo.cpp]: asking the wrong person

File: module/support/Gazebo/src/Gazebo.cpp [hunk 25]
```diff
@@ -0,0 +1,444 @@
...
+                if (pub.HasConnections())
+                    std::cout << "Has connections!" << std::endl;
+                else
+                    std::cout << "No connections..." << std::endl;
+            //NUClear::log(__LINE__);
+            }
+
+        });*/
+
+        on<Trigger<std::vector<ServoTarget>>, With<DarwinSensors>>().then(
```
[the remark this answers is not in the document]
Comment (Bidski) [reply module/support/Gazebo/src/Gazebo.cpp]: We do for this one

File: module/support/Gazebo/src/Gazebo.cpp [hunk 26]
```diff
@@ -0,0 +1,444 @@
...
+            });
+
+        on<Trigger<GazeboWorldCtrl>>().then([this](const GazeboWorldCtrl& command) {
+            ignition::msgs::StringMsg message;
+            message.set_data(command.command);
+            if (!this->worldPub.Publish(message)) std::cout << "Error publishing to world control topic!" << std::endl;
+        });
+    }
+
+    const ignition::msgs::StringMsg Gazebo::GenerateMsg() {
```
[the remark this answers is not in the document]
Comment (Bidski) [reply module/support/Gazebo/src/Gazebo.cpp]: not a variable?

File: nuclear/b.py (hunk 2 above)
Comment (Bidski) [reply nuclear/b.py]: yea, I think it was libzmq

File: puppet/manifests/nubots.pp (hunk 4 above)
Comment (Bidski) [reply puppet/manifests/nubots.pp]: dont get me started

File: puppet/manifests/nubots.pp (hunk 5 above)
Comment (Bidski) [reply puppet/manifests/nubots.pp]: no, its part of all the gazebo dependencies

File: puppet/modules/build_tools/manifests/init.pp (hunk 6 above)
Comment (Bidski) [reply puppet/modules/build_tools/manifests/init.pp]: dont get me started

File: puppet/modules/build_tools/manifests/init.pp (hunk 7 above)
Comment (Bidski) [reply puppet/modules/build_tools/manifests/init.pp]: not that i know of

File: puppet/modules/installer/manifests/init.pp (hunk 8 above)
Comment (Bidski) [reply puppet/modules/installer/manifests/init.pp]: dont get me started

File: shared/message/motion/ExecuteKick.proto (hunk 9 above)
Comment (Bidski) [reply shared/message/motion/ExecuteKick.proto]: you wrote it, you tell me!

File: shared/message/support/Gazebo/Gazebo.proto (hunk 10 above)
Comment (Bidski) [reply shared/message/support/Gazebo/Gazebo.proto]: havent a clue

File: shared/message/support/Gazebo/GazeboBallLocation.proto (hunk 12 above)
Comment (Bidski) [reply shared/message/support/Gazebo/GazeboBallLocation.proto]: you think we know what space this is in?

File: shared/message/support/Gazebo/GazeboBallLocation.proto (hunk 13 above)
Comment (Bidski) [reply shared/message/support/Gazebo/GazeboBallLocation.proto]: *shrugs*

File: shared/message/support/Gazebo/GazeboRobotLocation.proto (hunk 14 above)
Comment (Bidski) [reply shared/message/support/Gazebo/GazeboRobotLocation.proto]: you think we know what space this is in?

File: module/motion/IKKick/src/IKKick.cpp (hunk 23 above)
[the remark this answers is not in the document]
Comment (TrentHouliston) [reply module/motion/IKKick/src/IKKick.cpp]: I see that it used to be a random struct rather than a message. Should be put back as that shouldn't be a part of this PR

File: module/support/Gazebo/src/Gazebo.cpp [hunk 27]
```diff
@@ -0,0 +1,360 @@
...
+            discovery_node = std::make_unique<ignition::transport::MsgDiscovery>(p_uuid, gazebo_message_port);
+
+            msg_publisher = std::make_unique<ignition::transport::MessagePublisher>(
+                topic_ctrl, host_addr, ctrl_addr, p_uuid, n_uuid, "ADVERTISE", *ad_msg_opts);
+
+            msg_pub_world = std::make_unique<ignition::transport::MessagePublisher>(
+                topic_world_ctrl, host_addr, ctrl_addr, p_uuid, n_uuid, "ADVERTISE", *ad_msg_opts);
+
+            bool joint_status_cb =
+                joint_status->Subscribe<Gazebo, ignition::msgs::StringMsg>(topic_status, &Gazebo::JointStatusCB, this);
```
Comment (TrentHouliston) [inline module/support/Gazebo/src/Gazebo.cpp]: Second argument here is a protocol buffer type. So to swap over to using custom protocol buffers instead of using strings and parsing them, you can use this pattern.

```c++
joint_status->Subscribe<Gazebo, ::protobuf::message::input::SomeMessage>(topic_status, &Gazebo::JointStatus, this);
...
void Gazebo::JointStatus(const message::input::SomeMessage& msg) {
...
}
```
This should work as neutron stores all the messages in the same namespace with ::protobuf on the start, and those protocol buffers are castable to the Neutron types.

**Comment by ysims** [comment]
@wongjoel did you need this anymore or is everything you need on the newer branch now? 
I'm under the impression we will be sticking with Webots now. The league are using Webots for the forseeable future and I don't think it's worth maintaining two simulators.

**Comment by wongjoel** [comment]
> @wongjoel did you need this anymore or is everything you need on the newer branch now?
> I'm under the impression we will be sticking with Webots now. The league are using Webots for the forseeable future and I don't think it's worth maintaining two simulators.

I did a a quick read through and I believe we have everything we need for Webots from this PR in #540

**Comment by ysims** [comment]
Superseded by #540
